tdms/src/interpolate_Hfield.cpp

```diff
@@ -1,161 +1,231 @@
+
+We will use the notation (a_i,b_j,c_k) for the Yee cell indices, although bear in mind that this does not reflect the order the indices appear in the code.
+For example; if a = y, b = x, c = z, then the value of Ha at cell (a_i,b_j,c_k) is accessed via Ha[c_k][a_i][b_j], due to the interchanging of the x and y directions.
+Similarly; we will write Ha[a_i, b_j, c_k] to refer to the value of Ha associated to cell (a_i,b_j,c_k).
+
+Suppose now that we have selected cell a_i,b_j,c_k to interpolate to the centre of.
+We must interpolate in the b-direction, then c-direction, or vice-versa.
+For optimal accuracy we must determine the best interpolation scheme we can use in each direction at cell (a_i,b_j,c_k), and use the WORSE scheme second.
+
+Let us assume WLOG that the b-direction interpolation scheme (b_scheme) is inferior to that of the c-direction (c_scheme).
+We now need to interpolate Ha in the c-direction to obtain the value of Ha at the spatial positions (a_i, cell_b + Db, c_k) where
+b_j - b_scheme.index-1 + b_scheme.first_nonzero_coeff <= cell_b <= b_j - b_scheme.index-1 + b_scheme.last_nonzero_coeff.
 
-    // now we need to determine which of the y- and z-directions is optimal to perform the first interpolation in
-    // note: preference for avoiding the y-dimension because of possible 2D simulations
-    if (determine_second_dim(J, K, j, k)) {
-        // prefer y-dimension over z-dimension
-        interp_scheme y_scheme = determineInterpScheme(J, j);
-        if (better_scheme(y_scheme, CUBIC_INTERP_MIDDLE))
+    Let cell_b be one particular index in this range.
+    To apply c_scheme to obtain the value of Ha at (a_i, cell_b + Db, c_k), we require the values Ha[a_i, cell_b, cell_c] where
+    c_k - c_scheme-1 + c_scheme.first_nonzero_coeff <= cell_c <= c_k - c_scheme.index-1 + c_scheme.last_nonzero_coeff.
+    These values are fed to c_scheme.interpolate, which provides us with Ha at the spatial position (a_i, cell_b + Db, c_k).
+
+Now with approximations of Ha at each (a_i, cell_b + Db, c_k), we can pass this information to b_scheme.interpolate to recover the value of Ha at the centre of Yee cell (a_i, b_j, c_k).
+*/
+
+void interpolateTimeDomainHx(double ***Hxy, double ***Hxz, int i, int j, int k, int nJ, int nK, double *Hx)
+{
+    // Associations: a = x, b = y, c = z
+
+    // determine the z-direction scheme
+    const interpScheme &z_scheme = best_interp_scheme(nK, k);
+    // determine the y-direction scheme
+    const interpScheme &y_scheme = best_interp_scheme(nJ, j);
+
+    // this data will be passed to the second interpolation scheme
+    double data_for_second_scheme[8];
+    // this data will hold values for the interpolation in the first interpolation scheme
+    double data_for_first_scheme[8];
+
+    // which of z_scheme and y_scheme is WORSE? We will interpolate in this direction SECOND
+    if (z_scheme.is_better_than(y_scheme))
```

Comment:
The reason for the separation is because we have to read the data from different axes of the `E{x,y,z}` etc arrays. [There was this discussion](https://github.com/UCL/TDMS/pull/74#discussion_r951168214) about the same thing - there probably is a way to reduce the code repetition, but I can't see an immediate and clear way.

Reply:
Sorry for the duplicated comment!
But good to know I'm thinking along similar lines to Tom.